src/main/java/com/aws/greengrass/authorization/AuthorizationModule.java

```diff
@@ -42,17 +47,61 @@ public void addPermission(String destination, Permission permission) throws Auth
                 || Utils.isEmpty(permission.getOperation())) {
             throw new AuthorizationException("Invalid arguments");
         }
-        // resource as null is ok, but it should not be empty
         String resource = permission.getResource();
-        if (resource != null && Utils.isEmpty(resource)) {
-            throw new AuthorizationException("Resource cannot be empty");
-        }
+        String validatedResource = validateResource(resource);
         resourceAuthZCompleteMap.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                resource);
         rawResourceList.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                validatedResource);
+    }
+
+    /**
+     * Only allow '?' if it's escaped. You can only escape special characters ('*', '$', '?').
+     * Any occurrence of '${' is only valid if it holds a single valid special character ('*', '$', '?') inside it
+     * and ends with '}'. (eg: "${*}" is valid, "${c}" is invalid, "${c" is invalid, ${*bc} is invalid)
+     * @param resource resource to be parsed
+     */
+    private String validateResource(String resource) throws AuthorizationException {
```

Comment:
We allow `*` as a resource. Is that allowed by this?

Reply:
Yes